### PR TITLE
Validate only metrics for binary cross etropy

### PR DESCRIPTION
**Pull request by ybubnov**
This patch relaxes validation of the output vectors for categorical
cross entropy.

### CODE REVIEW

**Comment by ybubnov** [comment]
CC @Avcu

File: keras_metrics/metrics.py [hunk 1]
```diff
@@ -36,13 +37,13 @@ def _categorical(self, y_true, y_pred, dtype):
...
-        # When the shape had dimenstion 3 and more and the label is
+        # When the shape had dimension 3 and more and the label is
         # not specified, we should throw an error as long as calculated
         # metric is incorrect.
         _, labels = y_pred.shape
-        if labels == 2:
+        if self.loss == "binary_crossentropy" and labels == 2:
             return self._binary(y_true, y_pred, dtype, label=1)
-        elif labels > 2:
+        elif self.loss == "binary_crossentropy" and labels > 2:
```
Comment (Avcu) [inline keras_metrics/metrics.py]: I don't get it, can you explain in a detailed way. Why does someone wouldn't use loss function of mean_squared_error and want to see the recall and precision. And If you're right it should be
` elif self.loss == "categorical_crossentropy" and labels > 2:`

Comment (ybubnov) [reply keras_metrics/metrics.py]: The problem, that these metrics are unusable in case of categorical crossentropy loss. Even when there are 4 outputs, I could still calculate correct result, but this check throws an error in case, when more than 4 labels in the output shape.

Check out this example:
```py
import keras
import keras_metrics

num_classes = 10

model = models.Sequential()
model.add(keras.layers.Dense(512, activation="relu", input_dim=128))
model.add(keras.layers.Dense(128, activation="relu"))
model.add(keras.layers.Dense(num_classes, activation="softmax"))

model.compile(optimizer="nadam",
              loss="categorical_crossentropy",
              metrics=[keras_metrics.precision(), keras_metrics.recall()])
```

Comment (ybubnov) [reply keras_metrics/metrics.py]: Probably, this check is not required at all.

Comment (ybubnov) [reply keras_metrics/metrics.py]: I believe we could even define a separate package for metrics that are usable only for _binary crossentropy_.

Comment (Avcu) [reply keras_metrics/metrics.py]: Let's go back to old-school fashion. Let's say there are there are 3 classes and we have 9 data points. First 3 of them belong to the class 0, 3 of them belong to the class 1 and 3 of them belong to the 2, making our true_label = [0,0,0,1,1,1,2,2,2]. One has to use to_categorical() function for such case a problem definitely. And this function will turn the labels into following form.
true_label = 
[1,0,0;
1,0,0;
1,0,0;
0,1,0;
0,1,0;
0,1,0;
0,0,1;
0,0,1
0,0,1]
Okay, let's assume your model generates something like that 
prediction_label = 
[1,0,0; (correct)
0,1,0;  (false)
0,0,1;  (false)
1,0,0; (false)
0,1,0;  (correct)
0,0,1;  (false)
1,0,0; (false)
0,1,0;  (false)
0,0,1;  (correct)]
Then let's go over your code step by step that you think it calculates the correct result. So checking tp,tn,fp,fn  found by(I skipped the casting and rounding)

tp = K.sum(y_true * y_pred)
it'll make the tp=3 for our case(sample 1, sample 5 and sample 9 are correct)

neg_y_true = 1 - y_true
neg_y_pred = 1 - y_pred
tn = K.sum(neg_y_true * neg_y_pred)
it'll make the tn=12 (you can count if you want)

neg_y_true = 1 - y_true
fp = K.sum(neg_y_true * y_pred)
it'll make the fp=6 (you can count if you want)

neg_y_pred = 1 - y_pred
fn = K.sum(y_true * neg_y_pred)
it'll make the fn=6 (you can count if you want)

Lastly, let's sum all of them tp+tn+fp+fn=27. Does make me amaze? No. Because when there are two labels the number was twice of the batch size, now it is equal to three times of number of sample. Of course, this whole story doesn't make sense if you do not think that sum of these numbers does not necessarily should be equal to batch size. 
I do not know what that made you think that it calculates the correct number when there are 4 classes. Can you enlighten me?

Best.

Comment (Avcu) [reply keras_metrics/metrics.py]: I am repeating myself, but again these metrics defined for BINARY classification problems. The only way to use them on multi-class problem is that you interpret the result as if it's binary problem. Please check what I attached below.
http://scikit-learn.org/stable/auto_examples/model_selection/plot_precision_recall.html

You can see calculation of precision on binary, 
average_precision = average_precision_score(y_test, y_score)

and on multi-class problem
for i in range(n_classes):
    precision[i], recall[i], _ = precision_recall_curve(Y_test[:, i],
                                                        y_score[:, i])
    average_precision[i] = average_precision_score(Y_test[:, i], y_score[:, i])

Comment (ybubnov) [reply keras_metrics/metrics.py]: @Avcu, Sorry for a late response I was quite busy these days with my main job. After the reading your explanation now I clearly understand your point and where I was wrong. Thank you for your time! Closing this PR.
